[inequality] Review of lecture and incorporate updates

mmcky opened the pull request:
This PR is a full review of the **inequality** lecture

It incorporates the following feedback from RA reading groups (#380, #379, #297, #295)

## General

- [x]  review and add appropriate labels to figures x and y axis for figures
- [x]  ensure figures are captioned and numbered using `mystnb` metadata

## Code

- [x] write Lorenz Curve function in the lecture without the use of numba (reported to QuantEcon.py as well https://github.com/QuantEcon/QuantEcon.py/issues/721) to show students what that function does. Fixes #379 
- [x] write gini coefficient function in the lecture without the use of numba.

**Note:** this has been done but it gives the lecture a long run time when calculating `gini` coefficients for the US data across 50 years. 

- [x] switch to compute 1 year from the data (and then import the remaining years that are pre-computed from a data file) for comparing `wealth` and `income`. 
- [x] the lecture makes use of data from https://github.com/QuantEcon/high_dim_data but that is external to the lecture. It looks like that repository makes use of git-lfs for larger file storage so there may be a good reason to keep this as a data repository. I recommend we update its name to `data-lecture-python` or `lecture-python-data` perhaps? **This will be done in a separate PR due to complexity of changing names and the live site at the same time**
- [x] update this to use pandas to replace outliers as code is more intuitive.

````md
```{code-cell} ipython3
# use an average to replace an outlier in labor income gini
ginis_li_new = ginis_li
ginis_li_new[5] = (ginis_li[4] + ginis_li[6]) / 2
```
````
- [x] ~~find out why there is such a bit outlier in 1965 in the computation of `Gini` from `labour income` for the USA~~

jstac commented:
Thanks @emilsavalan , much appreciated!

There is a similar problem in the function `plot_value_function_seq` just above.  I'll fix that one.

jstac commented:
@mmcky , should this PR generate a view on the compiled page?

mmcky commented:
hey @jstac this is the old repo for python.quantecon.org

Changes should be made here:

https://github.com/quantecon/lecture-python.myst

One of the reasons I want to clean the RST based repos up and archive them.

jstac commented:
Thanks @mmcky .   I think it would be a good idea to archive them.

 @emilsavalan , sorry about the hassle. Would you like to put in a fresh PR in the new repo, which @mmcky listed above?  Or, if you prefer, I can make that change?

emilsavalan commented:
@jstac , it is okay. I commited changes, please check it.

jstac commented:
Thanks @emilsavalan .

I'll close this PR @mmcky and leave you to archive.  Thanks in advance.

mmcky committed:
[inquality] Review of lecture and incorporate updates

mmcky committed:
update plots

mmcky committed:
more plot updates

mmcky committed:
add quantecon back in for now re: gini_coefficient

mmcky commented:
idea from @jstac re: `gini coefficient` and high compute costs

> How about we show how to compute the gini on small simulated data and then download the gini time series from the world bank or world inequality database?
> That way we could get the gini time series for the UK and one or two Scandinavian countries as discussed in the meeting, and do a cross country comparison...

- [x] implement the above suggestion

mmcky committed:
@mmcky edits

mmcky committed:
review and minor edits

mmcky committed:
fix execution issue

mmcky committed:
minor updates to code output

mmcky committed:
updates

mmcky commented on lectures/inequality.md:
````diff
@@ -417,83 +616,98 @@ varlist = ['n_wealth',   # net wealth
+
 ```{code-cell} ipython3
-ginis_nw, ginis_ti, ginis_li = Ginis
+ginis = pd.read_csv("_static/lecture_specific/inequality/usa-gini-nwealth-tincome-lincome.csv", index_col='year')
````
- [ ] this would need to be updated once the `csv` file is published on GitHub

mmcky commented:
@jstac the last thing I need to do is compare `USA` with some other countries. This lecture is getting pretty long so I was wondering if we wanted to split this up into:

- `Inequality: How do we measure inequality?`
- `Inequality: Closer look at income and wealth in the USA`
- `Inequality: Cross country comparisons of income inequality`

jstac commented:
Thanks @mmcky .  Perhaps I'll let you make those last additions and then we can review and discuss how to slice and dice...

mmcky committed:
Compare inquality across countries

mmcky committed:
add plotly install for preview

mmcky committed:
supress mimetype warnings

mmcky committed:
TMP: temporarily disable warnings for preview

mmcky committed:
TMP: enable require.js

mmcky commented:
@jstac I am about to time out in 20min for this afternoon (home duties) but wanted to give you an update.

 I have added:

- cross-country comparisons in gini between US, UK and Norway
- a plot of GDP per capita vs gini (for US, UK and Normay) for all years (as a year-path scatter plot). I think it is an interesting graph as it plots a couple of dimensions but I need to declutter the labels by adding a 5 year skip on the year labelling. 

No need for any detailed review but I would be interested your thoughts on the additional charts in the new sections. 

Preview: https://65ea94ae2e9af43492501c44--taupe-gaufre-c4e660.netlify.app/inequality.html

I need to:

- [x] add a 5 year skip to the year labels in the gdp per capita vs gini figure
- [x] update new additions to make them quantecon style-guide compliant (label figures etc.)
- [x] review all issues linked in the main comments section to make sure we address all comments
- [x] finalise narrative additions to make the flow of the lecture more cohesive with the new sections
- [x] proof-read.
- [x] hand over to @jstac

jstac commented on lectures/inequality.md:
````diff
@@ -124,18 +123,73 @@ The meaning of the statement $y = L(x)$ is that the lowest $(100
+of people and the cumulative share of income (or wealth).
+
+```{code-cell} ipython3
+:tags: [hide-input]
````
I vote for un-hiding this.  It's nice clear code and not too  long.

jstac commented on lectures/inequality.md:
```diff
@@ -280,44 +328,49 @@ Lorenz curve.
 The Gini coefficient is closely related to the Lorenz curve.
 
 In fact, it can be shown that its value is twice the area between the line of
 equality and the Lorenz curve (e.g., the shaded area in the following Figure below).
```
(e.g., ...the following Figure below) -> , as illustrated in "Figure xx"

jstac commented on lectures/inequality.md:
````diff
@@ -280,44 +328,49 @@ Lorenz curve.
+```{note}
+Another way to think of the gini coefficient is the area between the 45-degree line of 
+perfect equality and the Lorenz curve minus the area below the Lorenz curve devided by
+the total area below the 45-degree line. 
````
Is this correct?  It sounds surprising.  Could we please have a reference?

mmcky replied:
thanks @jstac I have clarified in [de796df](https://github.com/QuantEcon/lecture-python-intro/pull/384/commits/de796df498bd29fb8fd7440341a4ed66f2806ca9)

jstac commented on lectures/inequality.md:
```diff
@@ -327,7 +380,7 @@ ax.fill_between(f_vals, l_vals, f_vals, alpha=0.06)
 ax.set_xlim((0, 1))
 
-ax.text(0.04, 0.5, r'$G = 2 \times$ shaded area', fontsize=12)
+ax.text(0.04, 0.5, r'$G = 2 \times$ shaded area')
```
In line 388 below, please drop "dynamics" from the title.

jstac commented on lectures/inequality.md:
````diff
@@ -362,53 +448,166 @@ ginis = []
+Another, and often useful way to find series ID, is to use the [World Bank data portal](https://data.worldbank.org) and then use `wbgapi` to fetch the data.
+```
+
+Let us fetch the data for the USA.
````
Is this the Gini for wealth or income?

mmcky replied:
income and updated this morning

jstac commented on lectures/inequality.md:
```diff
@@ -362,53 +448,166 @@ ginis = []
+y = data_usa.dropna().values
+plt.scatter(x,y)
+plt.plot(x1, a1*x1+b1, 'r-')
+plt.plot(x2, a2*x2+b2, 'y-')
```
Use default matplotlib colors?

jstac commented on lectures/inequality.md:
```diff
@@ -362,53 +448,166 @@ ginis = []
+plt.scatter(x,y)
+plt.plot(x1, a1*x1+b1, 'r-')
+plt.plot(x2, a2*x2+b2, 'y-')
+plt.title("USA gini coefficient dynamics")
```
Gini should be capitalized, US rather than USA.

jstac commented on lectures/inequality.md:
```diff
@@ -362,53 +448,166 @@ ginis = []
+plt.legend(['Gini coefficient', 'Trend (before 1981)', 'Trend (after 1981)'])
+plt.ylim(25,45)
+plt.ylabel("Gini coefficient")
+plt.xlabel("Year")
```
"year" (as per style manual)

jstac commented on lectures/inequality.md:
```diff
@@ -362,53 +448,166 @@ ginis = []
+plt.plot(x2, a2*x2+b2, 'y-')
+plt.title("USA gini coefficient dynamics")
+plt.legend(['Gini coefficient', 'Trend (before 1981)', 'Trend (after 1981)'])
+plt.ylim(25,45)
```
There's a lot of whitespace on this fig too...

jstac commented on lectures/inequality.md:
````diff
@@ -362,53 +448,166 @@ ginis = []
+
+```{code-cell} ipython3
+gini_all.plot(kind="hist", title="Gini coefficient");
+```
````
The bar graph might look better with a few more bins.  And maybe outlines to show the bars?  I don't love the default bar chart in matplotlib...

Lowercase plot labels needed

jstac commented on lectures/inequality.md:
```diff
@@ -362,53 +448,166 @@ ginis = []
+
+We can use the data collected above {ref}`survey of consumer finances <data:survey-consumer-finance>` to look at the gini coefficient when using income when compared to wealth data. 
+
+Let's compute the gin coefficient for net wealth, total income, and labour income for the most recent year in our sample.
```
gin -> Gini

jstac commented on lectures/inequality.md:
```diff
@@ -417,83 +616,98 @@ varlist = ['n_wealth',   # net wealth
 
 Let's plot the Gini coefficients for net wealth, labor income and total income.
 
+Looking at each data series we see an outlier in gini coefficient computed for 1965. 
```
gini -> Gini

mmcky replied:
this has been done in today's narrative updates 👍

jstac commented on lectures/inequality.md:
````diff
@@ -417,83 +616,98 @@ varlist = ['n_wealth',   # net wealth
+```
+
+```{code-cell} ipython3
+ginis["l_income"].plot()
````
Add labels

jstac commented:
Thanks @mmcky , nice work!  Minor comments above.

The series for Norway is pretty short --- is there a similar country we could substitute in?  We'll have to do a bit better somehow...

mmcky commented:
> Thanks @mmcky , nice work!  Minor comments above.
> 
> 
> 
> The series for Norway is pretty short --- is there a similar country we could substitute in?  We'll have to do a bit better somehow...

Thanks John - I'll work then to incorporate and update the narrative text etc. 

There are additional years for Norway but being dropped because not contiguous data points. Will fix that. You can see some earlier Norway data points in the plotly graph.

mmcky committed:
Merge branch 'main' into review-inequality

mmcky committed:
@mmcky review and proof-read

mmcky committed:
review of executable version in jupyter lab

mmcky committed:
address @jstac comment 1

mmcky committed:
drop dynamics from title of section

mmcky committed:
use matplotlib default cycler

mmcky committed:
update gini to Gini

mmcky committed:
set x and y labels

mmcky committed:
review in jupyter-lab

mmcky committed:
tweaks from review of preview

mmcky committed:
remove incorrect statement

mmcky committed:
minor edit

mmcky committed:
improve narrative

mmcky commented:
@jstac I have just changed the status of this from `in-work` to `review`. 

I think this is in a good state for review and comment.

jstac commented:
IMO, the tip here grabs most of the readers attention, even though the definition is much more important.

![image](https://github.com/QuantEcon/lecture-python-intro/assets/3887684/2c3b8451-9bf6-463b-9771-5a340707a8ee)

jstac commented:
Regarding the last point, perhaps it would be best to use the definition environment from sphinx proof for the Lorenze curve and drop the tip (just put the sentence in brackets.?

If we do this, the Gini coefficient should also be defined in the same environment.

jstac commented:
As mentioned before, dropping `ax.set_ylim(0, data_usa.max() + 5)` in the first Gini dynamics plot will make it look nicer.  There's too much whitespace ATM.

jstac commented:
In the same figure, on the x-axis, 1980 instead of YR1980 etc.

jstac commented:
"does not have significant variation in the full range from 0 to 100." -> "trended upward from 1980 to 2020 and then dropped slightly following the COVID pandemic".

mmcky commented:
> As mentioned before, dropping `ax.set_ylim(0, data_usa.max() + 5)` in the first Gini dynamics plot will make it look nicer. There's too much whitespace ATM.

@jstac I updated the plot labelled `gini_usa_trend` to focus more intensely on the variation within the data. I left this plot to give context (of variance) based on the feedback from the reading group and made the two points below it along with the narrative talking about limited variation in the Gini coefficient. 

This was the reason for the proposed white space in context with the points below the graph. 

```
1. moves slowly over time, and 
2. does not have significant variation in the full range from 0 to 100.
```

The text then goes on to talk about vairation in Gini (across countries and time) then zoom's in on the US data to highlight the variation in a stepped fashion. 

**Update:** I have added your sentence as the first point. 

```md
As can be seen in {numref}`gini_usa1` the Gini coefficient:

1. trended upward from 1980 to 2020 and then dropped slightly following the COVID pandemic
1. moves slowly over time
3. does not have significant variation in the full range from 0 to 100
```

mmcky committed:
incorporate some of @jstac comments

mmcky commented:
> In the same figure, on the x-axis, 1980 instead of YR1980 etc.

I have gone ahead and fixed this on data import -- which fixes up the plots. The world bank database doesn't always return in a shape or a type that we want.

jstac commented:
'Trend (before 1981)' -> 'trend (before 1981)'

On the y-axis 'Frequency' -> 'frequency'

(As per the style manual.)

jstac commented:
Remove output `<Axes: ylabel='gini coefficient'>`

Above 'We see that Norway' on the y-axis, gini coefficient -> Gini coefficient.

I prefer without the 'economy' heading for the labels --- my philosophy is cut everything you can possibly cut without losing clarity, to focus attention on what's important.

jstac commented:
Hey @mmcky , great work!   Very minor criticism above -- thanks for your patience :-)

I love the plot with Norway, the US and GBR.  Well done!  It's really instructive.  I'll have to write something about Reagan and Thatcher -- I can't resist.

If you're willing, I'd like the following additional changes:

- All figures shifted to a figure environment
- For the plots with Gini coefficients, clarify whether it's a wealth Gini or an income Gini or both in every case (some are missing this information)

I would be inclined to merge current figs 5.6 and 5.7 -- and even drop the trend lines.  Less is more and we only want to convey one simple message -- the Gini fell until 1980 and then rose.

mmcky commented:
- [x] fix display formulae in prf:definition directives

<img width="831" alt="Screenshot 2024-03-12 at 4 48 27 PM" src="https://github.com/QuantEcon/lecture-python-intro/assets/8263752/28d8a5a5-06b4-4058-aebc-8282426b90cf">

mmcky commented:
@jstac do you think we should make the direction of the Gini coefficient easier to understand on plots with an additional axis such as

![image](https://github.com/QuantEcon/lecture-python-intro/assets/8263752/b112cc7c-f12a-492e-a7ec-c8c43d160794)

jstac commented:
I hate to say this but I'm not sure that figure should make the final cut.  Does it have a clear economic message?

mmcky commented:
> I hate to say this but I'm not sure that figure should make the final cut. Does it have a clear economic message?

No worries either way re: final cut. 

My thoughts on this was to show the relationship (or lack there of between Gini and GDP per capita) over time without dwelling on the issues (as a main point of discussion) of causality and growth vs. inequality. My thinking of this graph is that it highlights key structural differences between `Norway` and the `US` economies showing steady progress in GDP per capita over time and very different Gini dynamics. I found it interesting that the UK became a lot more like the US from the early 1980s in terms of the `Gini` coefficient.

jstac commented:
> My thinking of this graph is that it highlights key structural differences between Norway and the US economies showing steady progress in GDP per capita over time and very different Gini dynamics. I found it interesting that the UK became a lot more like the US from the early 1980s in terms of the Gini coefficient.

True.  And it's a very nice looking figure :-)

mmcky commented:
An alternative is to look at Happiness and Gini (as per suggestion in #380)

mmcky committed:
fix equation in prf:definition

jstac commented:
@mmcky Please let me know when you have addressed my requests above.  I think the figure you've added on the Gini coefficient and GDP dynamics is fine as is.  Let's keep it.

mmcky committed:
address remaining @jstac direct comments

mmcky committed:
Merge branch 'main' into review-inequality

mmcky committed:
save branch conflict

mmcky committed:
@mmcky final review

mmcky commented:
@jstac I did another review and checked the above comments. I think all changes are now incorporated except for:

- flow of graphs re: US and Gini (https://github.com/QuantEcon/lecture-python-intro/pull/384#issuecomment-1990258661) to emphasise and discuss the variation of Gini. I have removed trend lines as they were unnecessary. 
- I have not incorporated Happiness vs. Gini (but I can do this once this is merged). 

These are all just suggestions. @jstac please feel free to use as you see fit. Once you edit I will review in detail to better match your style in edits. 

- [x] @mmcky to double check `html` output once preview is available

mmcky committed:
update figures with mystnb figure settings

mmcky committed:
review from html

mmcky commented:
@jstac in [3b65128](https://github.com/QuantEcon/lecture-python-intro/pull/384/commits/3b65128e433a313fcadeb2ee853275b420d78af9) I have removed the `mystnb` figure wrapper around the plotly figure. There is a current bug in that it labels the figure twice (as plotly returns two objects) in the output stream. That is something we will need to fix in jupyter-book

mmcky committed:
fix line continuation

mmcky commented:
@jstac I think this is now ready for review.

jstac commented:
The link `https://ourworldindata.org/what-is-the-gini-coefficient` is broken -- due to extra `]` at the end?

jstac commented:
It occurred to me that the computation of the Gini coefficient could be vectorized:

```python
def gini(y):
    n = len(y)
    y_1 = np.reshape(y, (n, 1))
    y_2 = np.reshape(y, (1, n))
    g_sum = np.sum(np.abs(y_1 - y_2))
    return g_sum / (2 * n * np.sum(y))
```

If that's not needed, it could still be an exercise asking the reader to produce a faster NumPy version that doesn't use loops and checking that it produces the same output.

jstac commented:
Please eliminate fig 5.6 --- the lecture is already long and I don't think it adds much value over 5.8.  In 5.8, please add "income" to the figure label.

Please change 

"Gini coefficient for US data (income)
Now let’s look at the Gini coefficient using US data."

to 

"Gini coefficient for income (US data)"
Let's look at the Gini coefficient for the distribution of income in the US."

jstac commented:
I think fig 5.10 is a little confusing when it follows on from fig 5.8.  Why does the Gini coefficient for total income look so different?  Why are we looking at the income Gini twice?  I suggest we drop the comparison and this data, and change the title and start of section 5.3.4 to 

"Gini coefficient for wealth (US data)"

In the previous section we looked at the Gini coefficient for income using US data.  

Now let's look at the Gini coefficient for the distribution of wealth."

Please then delete

"As we have discussed the Gini coefficient can also be computed over different distributions such as income and wealth."

and suitably modify

"We can use the data collected above [survey of consumer finances](https://65fb98df7cd11a6190c5b679--taupe-gaufre-c4e660.netlify.app/inequality#data-survey-consumer-finance) to look at the Gini coefficient when using income when compared to wealth data."

jstac commented:
Please add "for income" to the figure titles for figs 5.12 and 5.13.

jstac commented:
Hey @mmcky , many thanks.  The lecture is looking very good.  I have some more suggestions above.  Sorry that they haven't come all at once.  It's mainly a case of clarifying and cutting to emphasize the clearest results.

mmcky commented:
> It occurred to me that the computation of the Gini coefficient could be vectorized:
> 
> ```python
> def gini(y):
>     n = len(y)
>     y_1 = np.reshape(y, (n, 1))
>     y_2 = np.reshape(y, (1, n))
>     g_sum = np.sum(np.abs(y_1 - y_2))
>     return g_sum / (2 * n * np.sum(y))
> ```
> 
> If that's not needed, it could still be an exercise asking the reader to produce a faster NumPy version that doesn't use loops and checking that it produces the same output.

Thanks @jstac I think keeping the code simpler is a nice thing -- but I love the idea of making this an exercise.

mmcky committed:
Incorporate @jstac feedback and comments

mmcky commented:
@jstac I wrote an exercise but it had a long run time when using `us data`. So I want to think about making it better using simulation data. I have opened https://github.com/QuantEcon/lecture-python-intro/issues/410 to track the addition so it doesn't have to hold this PR up.

jstac commented:
I'm surprised this is green.  The build is failing after the code cell starting `# Fetch gini data for all countries`

mmcky commented:
Well that is fascinating. It is working locally -- so I suspect the wb database thought some ip was requested too much data and returned the 400. But fascinating that an API Error doesn't get recongised as a python error.

mmcky committed:
reinstate hardline error on warning for html build on CI

mmcky commented:
@jstac the new run worked OK

https://66020e1f7cfe606404b17ca6--taupe-gaufre-c4e660.netlify.app/inequality

I have re-enabled build failure on warnings which takes an extra hardline on these types of warnings and failures so any build failures will prevent the CI from running fully.

mmcky committed:
Fix: duplicate label

mmcky committed:
add ignore on myst.domain warnings

jstac commented:
Thanks @mmcky !  It's looking great.  Minor comments:

- [x] This sentence needs to be adjusted: "Let us zoom on the US data so we can more clearly observe trends."
- [x]  I wonder if we should remove the discussion starting "Looking at each data series we see an outlier in Gini coefficient computed for 1965 for labour income." It's not clear why the labor income distribution is so different to the income distribution in fig 5.7.  I feel like we either need to explain that or get rid of the labor income / wealth comparison.
- [x]  "The wealth time series exhibits a strong U-shape." should probably be after fig 5.8.
- [x]  "As we saw earlier in this lecture " -> "Earlier in this lecture"
- [x] "western economies" -> "Western economies"

mmcky commented:
> I wonder if we should remove the discussion starting "Looking at each data series we see an outlier in Gini coefficient computed for 1965 for labour income." It's not clear why the labor income distribution is so different to the income distribution in fig 5.7. I feel like we either need to explain that or get rid of the labor income / wealth comparison.

@jstac I would agree with this comment - I think we should get rid of the labour income comparison. I don't have a definitive reason why the series looks so different to Fig 5.7 (income). https://github.com/QuantEcon/lecture-python-intro/issues/412

mmcky committed:
address @jstac feedback

mmcky commented:
thanks @jstac for your review and comments. I have addressed the latest round of feedback.

jstac committed:
misc

jstac commented:
Many thanks @mmcky .  The lecture looks great.  I've pushed some minor edits.  If it builds correctly then please go ahead and merge.

mmcky commented:
thanks @jstac

mmcky commented:
thanks @jstac for all your excellent comments and feedback. Will incorporate lessons learnt re: style in the next review.
